provision/filter_plugins/aws_inventory.py

```diff
@@ -0,0 +1,110 @@
+import StringIO
+import pprint 
+def pop_items(item_list,count):
+    pop_list = []
+    for i in range(0,count):
+        pop_item = item_list.pop()
+        pop_list.append(pop_item)
+    return pop_list,item_list
+
+def get_host_ips(topo):
+    host_public_ips = []
+    for group in topo['aws_ec2_res']:
+        #host_public_ips.append(str(group['server']['public_v4']))
+        for instance in group['instances']:
+            host_public_ips.append(str(instance['public_dns_name']))
+    return host_public_ips
+
+def get_layout_hosts(inv):
+    return inv['hosts'].keys()
+
+def get_layout_host_groups(inv):
+    return inv['host_groups'].keys()
+
+def add_sections(config, section_list):
+    for section in section_list:
+        config.add_section(section)
+    return config
+
+def set_children(config, inv):
+    for host_group in inv['host_groups']:
+        if "children" in inv['host_groups'][host_group]:
+            config.add_section(host_group+":"+"children")
+            for child in inv['host_groups'][host_group]['children']:
+                config.set(host_group+":"+"children",child)
+    return config    
+
+def set_vars(config,inv):
+    for host_group in inv['host_groups']:
+        if "vars" in inv['host_groups'][host_group]:
+            config.add_section(host_group+":"+"vars")
+            for var in inv['host_groups'][host_group]['vars']:
+                config.set(host_group+":"+"vars",var, inv['host_groups'][host_group]['vars'][var])
+    return config
+
+def add_ips_to_groups(config,inven_hosts,layout):
+    # create a ip to host mapping based on count
+    ip_to_host = {}
+    for host_name in layout['hosts']:
+        count = layout['hosts'][host_name]['count']
+        host_list = []
+        for i in range(0,count):
+            item = inven_hosts.pop()
+            host_list.append(item)
+        ip_to_host[host_name] = host_list
+    # add ips to the host groups in inventory  
+    for host_name in layout['hosts']:
+        host_ips = ip_to_host[host_name]
+        for ip in host_ips:
+            for host_group in layout['hosts'][host_name]['host_groups']:
+                config.set(host_group,ip)
+    return config
+
+def add_common_vars(config, host_groups, layout):
+    common_vars = layout['vars']
+    for group in host_groups:
+        items = dict(config.items(group)).keys()
+        config.remove_section(group)
+        config.add_section(group)
+        for item in items:
+            host_string = item
+            for var in common_vars:
+                if common_vars[var] == "__IP__":
+                   host_string += " "+ var +"="+item+" "
+            config.set(group,host_string)
+    return config
+    
+def aws_inventory(topo, layout):
+    inventory = ConfigParser(allow_no_value=True)
+    no_of_groups = len(topo['aws_ec2_res'])
+    layout_hosts = get_layout_hosts(layout)
+    inven_hosts = get_host_ips(topo)
+    if len(inven_hosts) != len(layout_hosts):
```

Comment:
Initially, the idea was that if we had >1 clouds involved, we wouldn't track the hosts vs inventory nodes here. Instead, we'd track it in a generic place. Essentially, running through the inventory layout until we run out of items is likely going to be the way forward. We only care that there are enough hosts to handle whatever is in the layout. 

For duffy, this was simple, since it'll never really be used with any of the other cloud providers. However, consider this scenario:

Let's you have say 3 openstack nodes, 1 aws node, and 1 gce node. 
For the inventory layout, you have a need of four (4) nodes to generate an inventory.

Based upon this, you'd want to have a loop that goes through the nodes in some order. I'd suspect based upon a variable (like I have above 'inventory_playbooks' loop), which then would process the inventories based upon the order of the list. If the list were as follows:

gce, openstack, aws

And get inventory output files like:

inventory/gce.inventory (representing one node)
inventory/openstack.inventory (representing 3 nodes)

Even though linch-pin spins up 5 nodes, the aws node is not used in the inventory going forward.

This is completely acceptable. However, if you had a need for 6 nodes, linch-pin should at least alert you (or fail, depending on what we decide here) that you don't have enough nodes to generate all of the inventories correctly.

Thoughts?